player/javascript.c

```diff
@@ -0,0 +1,1237 @@
+static void push_failure(js_State *J, const char *str)
+{
+    set_last_error(jctx(J), 1, str);
+    js_pushundefined(J);
+}
+
+// Most of the scripting APIs are either sending some values and getting status
+// code in return, or requesting some value while providing a default in case an
+// error happened. These simplify the C code for that and always set last_error.
+
+static void push_status(js_State *J, int err)
+{
+    if (err >= 0) {
+        push_success(J);
+    } else {
+        push_failure(J, mpv_error_string(err));
+    }
+}
+
+ // If err is success then return 0, else push the item at def and return 1
+static bool pushed_error(js_State *J, int err, int def)
+{
+    bool iserr = err < 0;
+    set_last_error(jctx(J), iserr, iserr ? mpv_error_string(err) : NULL);
+    if (!iserr)
+        return false;
+
+    js_copy(J, def);
+    return true;
+}
+
+/**********************************************************************
+ *  Autowash - care-free resource deallocation on vm errors, and otherwise
+ *********************************************************************/
+// - Autowash (aw) functions are called with a pre-allocated and zeroed autowash
+//   object argument which is also a talloc context. The object's members and
+//   the object itself are freed after the function exits - either normally or
+//   because it threw an error, on the latter case it then re-throws the error.
+//
+//   Functions with autowash* argument should be inserted as js C functions with
+//   aw_newcfunction and used normally, or called from C with the AW macro.
+
+typedef struct autowash {
+    char *ma;  // mpv alloc -> mpv_free(..)
+    mpv_node result_node;  // a node obtained via an mpv API
+    FILE *f;   // fclose
+    DIR *dir;  // closedir
+} autowash;
+
+static autowash *new_autowash(void)
+{
+    assert(MPV_FORMAT_NONE == (mpv_format){0});  // else we misdetect on delete
+    return (autowash*)talloc_zero(NULL, autowash);
+}
+
+static void delete_autowash(autowash *aw) {
+    if (aw->ma)
+        mpv_free(aw->ma);
+    // "If node->format is MPV_FORMAT_NONE, this call does nothing." (client.h)
+    if (aw->result_node.format != MPV_FORMAT_NONE)
+        mpv_free_node_contents(&aw->result_node);
+    if (aw->f)
+        fclose(aw->f);
+    if (aw->dir)
+        closedir(aw->dir);
+
+    talloc_free(aw);
+}
+
+// Call a C function with autowash cleanup. The callee prototype should include
+// an autowash* argument, expr should use "aw" where the callee expects it,
+// and "J" is assumed. E.g. int foo(js_State *J) { AW(bar(J, "hello", aw)); }
+// This can throw, but if it does, then it's after the cleanup.
+// Macro because the callee could have any prototype.
+#define AW(expr) do {\
+    autowash *aw = new_autowash();\
+    if (js_try(J)) {\
+        delete_autowash(aw);\
+        js_throw(J);\
+    }\
+    expr;\
+    js_endtry(J);\
+    delete_autowash(aw);\
+} while (0)
```

Comment:
A `js_pcall` would be welcome, because it isolates the tricky setjmp handling into a separate function, and does not bother every caller with those weird semantics for local variables.

So yeah, I'm asking that you change this. There aren't many AW() users anyway.